File: APCSCI-Week5/H2.cpp
#include<bits/stdc++.h>
using namespace std;
typedef long long int ll;

bool bfscheck(char a, char b, unordered_map<char, vector<char>> & gr){
    queue<char> q;
    
}

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);
    int n;
    cin >> n;
    string alpha = "abcdefghijklmnopqrstuvwxyz";
    vector<string> words;
    for (int i = 0; i < n; i++){
        string in;
        cin >> in;
        words.push_back(in);
    }

    unordered_map<char, vector<char>> gr;
    for (char c : alpha){
        gr[c];
    }

    queue<vector<int>> query;
    vector<int> tmp = {0, n-1, 0};
    query.push(tmp);
    vector<char> firstchars;
    
    while (!query.empty()){
        int l, r, c;
        
        tmp = query.front(); query.pop();
        l = tmp[0];
        r = tmp[1];
        c = tmp[2];
        vector<int> toq = {l, 0, c+1};
        bool isfirst = false;

        for (int i = l; i < l+r; i++){
            string fi = words[i];
            string se = words[i+1];
            if (se.size() <= c && fi.size() > c){
                cout << "Impossible" << endl;
                return 0;
            }
            if (fi.size() <= c){
                continue;
            }

            if (fi[c] == se[c]){
                toq[0] ++;
                continue;
            }
            else if (fi[c] != se[c]){
                if (toq[1] > 0){
                    query.push(toq);
                }
                toq = {i+1, 0, c+1};
            }

            if (bfscheck(se[c], fi[c], gr)){
                gr[fi[c]].push_back(se[c]);
                isfirst = true;
            }
        }
    }
   
    return 0;
}

File: APCSCI-Week5/H2.py
from collections import deque

def ex():
    print("Impossible")
    exit()

def bfscheck(a, b, gr):
    # check if a will go to b. if it will, return false
    q = deque([a])
    vis = {chr(i):0 for i in range(97, 123)}
    vis[a] = 1
    while q:
        cur = q.popleft()
        for next in gr[cur]:
            if next == b:
                return False
            if not vis[next]:
                q.append(next)
                vis[next] = 1
    return True

def dfs(cur, ans, gr, visdfs):
    if visdfs[cur]:
        return
    stack = [cur]
    visdfs[cur] = 1
    while stack:
        #print(stack)
        v = stack.pop()
        ans.append(v)
        for next in gr[v]:
            if not visdfs[next]:
                visdfs[next] = 1
                stack.append(next)

def main():
    n = int(input())
    alpha = list("abcdefghijklmnopqrstuvwxyz")
    aset = set(alpha)

    words = [input() for i in range(n)]
    gr = {chr(i):[] for i in range(97, 123)}

    query = deque([[0, n-1, 0]])
    firstchars = []
    while query:
        l, r, c = query.popleft()
        toq = [l, 0]
        #print(l, r, c)
        isfirst = False
        
        for i in range(l, l+r):
            fi = words[i]
            se = words[i+1]
            #print(fi, se)
            if len(se) <= c and len(fi) > c:
                ex()
            if len(fi) <= c:
                continue

            if fi[c] == se[c]:
                toq[1] += 1
                continue
            elif fi[c] != se[c]:
                if toq[1] > 0:
                    query.append(toq + [c+1])
                toq = [i+1, 0]
            
            if bfscheck(se[c], fi[c], gr):
                gr[fi[c]].append(se[c])
                isfirst = True
            else:
                ex()
        if toq[1] > 0:
            query.append(toq + [c+1])

        if isfirst:
            firstchars.append(words[l][c])
        

        
    # for k, v in gr.items():
    #     print(k, v)

    ans = []
    #print(firstchars)
    visdfs = {chr(i):0 for i in range(97, 123)}
    for char in firstchars:
        dfs(char, ans, gr, visdfs)

    #print(*ans, sep='')
    anset = set(ans)
    for i in alpha:
        if i not in anset:
            ans.append(i)

    print(*ans, sep='')

main()